genie-agent/src/main/java/com/netflix/genie/agent/execution/services/DownloadManifest.java

```diff
@@ -0,0 +1,106 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.net.URI;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Manifest containing source (URIs) and their expected destination on disk after download.
+ * This abstraction is used by different services to set up the job working directory.
+ *
+ * Currently, source URIs must be unique. I.e. it is not possible to have a manifest with the same source
+ * URI and two or more target destinations.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+public interface DownloadManifest {
+
+    /**
+     * Get the set of files representing expected target location for files to be downloaded.
+     * @return an immutable set of files objects
+     */
+    Set<File> getTargetFiles();
+
+    /**
+     * Get the set of directories that contain all the target files.
+     * @return an immutable set of file objects
+     */
+    Set<File> getTargetDirectories();
+
+    /**
+     * Get the set of URI s (source files) that were added to the manifest.
+     * @return an immutable set of URIs
+     */
+    Set<URI> getSourceFileUris();
+
+    /**
+     * Get the map entries: URI source -> File target.
+     * @return an immutable set of map entries
+     */
+    Set<Map.Entry<URI, File>> getEntries();
```

Comment:
I didn't want to expose that it's a map internally, because the implication of that is that you cannot have the same URL downloaded to two different locations.
I think right now it's ok to have that limitation, but if we ever need to change, exposing this set is more future-proof.

Reply:
But it's a set of `Map.Entry` so you're still exposing internal representation. I don't think it's a huge deal either way just seemed like a useless API if you just have the getter for the map if someone wants the entry set they can get it directly.

Reply:
Doh! It should be a set of `Pair<URL, File>` or something to properly do the thing I was saying above.